### PR TITLE
Cleanup more User32 interop

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Cleanup more User32 interop

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3174)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MessageBeep.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class User32
+    {
+        [DllImport(Libraries.User32, ExactSpelling = true)]
+        public static extern BOOL MessageBeep(uint type);
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MessageBeep.cs:13]: It is missing `SetLastError`. Also `type` is `MB` enum (though we have more items in it)
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messagebeep

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetPhysicalCursorPos.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+using System.Drawing;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class User32
+    {
+        [DllImport(Libraries.User32, ExactSpelling = true)]
+        public static extern BOOL GetPhysicalCursorPos(ref Point lpPoint);
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetPhysicalCursorPos.cs:13]: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getphysicalcursorpos
`SetLastError` is missing

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+using System;
+using System.Drawing;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class User32
+    {
+        [DllImport(Libraries.User32, ExactSpelling = true, CharSet = CharSet.Auto)]
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs:13]: Why do we need `CharSet`?

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs [hunk 4]
```diff
@@ -2358,9 +2358,10 @@ private void NativeClear()
         private string NativeGetItemText(int index)
         {
             int len = unchecked((int)(long)SendMessageW(this, (WM)CB.GETLBTEXTLEN, (IntPtr)index));
-            StringBuilder sb = new StringBuilder(len + 1);
-            UnsafeNativeMethods.SendMessage(new HandleRef(this, Handle), (int)CB.GETLBTEXT, index, sb);
-            return sb.ToString();
+            return string.Create(len + 1, this, (span, comboBox) =>
+            {
+                SendMessageW(comboBox, (WM)CB.GETLBTEXT, (IntPtr)index, ref span[0]);
+            });
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs:2364]: Nice 👍

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MessageBeep.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MessageBeep.cs:13]: I thought we are ignoring `SetLastError` if we don't check it. I remember some comments too about it having a perf impact

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetPhysicalCursorPos.cs (hunk 2 above)
Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetPhysicalCursorPos.cs:13]: see above

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs (hunk 3 above)
Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs:13]: fixed

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MessageBeep.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.MessageBeep.cs]: > I thought we are ignoring `SetLastError` if we don't check it. I remember some comments too about it having a perf impact

Funnily enough few days ago @danmosemsft asked why we don't have it on all interops that provide it - https://github.com/dotnet/winforms/issues/3151
My (limited) understanding of the interop layer the perf impact isn't substantial for us to really care. But the absence of `SetLastError` makes people question whether the interops are defined and used correctly.

**Comment by hughbe** [comment]
Updated to add `SetLastError=true`

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+using System;
+using System.Drawing;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class User32
+    {
+        [DllImport(Libraries.User32, ExactSpelling = true)]
+        public unsafe static extern uint GetDlgItemInt(IntPtr hDlg, int nIDDlgItem, BOOL* lpTranslated, BOOL bSigned);
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs:14]: SetLastError is missing 
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getdlgitemint#return-value

Comment (hughbe) [reply src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetDlgItemInt.cs:14]: done

**Comment by RussKie** [comment]
Thank you

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs (hunk 4 above)
Comment (kpreisser) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs:2364]: Note that this change causes an additional NUL character to be returned in `NativeGetItemText`.
For example, if a combobox item string is `"ABC"`, before this change `NativeGetItemText` returned `"ABC"` (3 chars), but now it returns `"ABC\0"` (4 chars).

Was this intended?

Comment (kpreisser) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs:2364]: Note that this change (like the one in `ListBox`) causes an additional NUL character to be returned in `NativeGetItemText`.
For example, if a combobox item string is `"ABC"`, before this change `NativeGetItemText` returned `"ABC"` (3 chars), but now it returns `"ABC\0"` (4 chars).

Was this intended?

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs:2364]: @hughbe bump

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs:2364]: You're right that it now returns "ABC\0", which is unintended.

So I wrote some more tests in #3107 that exercise this code path and the tests passed on .NET Framework and .NET Core. It seems like we also already had tests testing this code path as well.

I was a little confused why the tests passed until I realised that the only place where this method was called was in code like

````cs
if (string.Compare(owner.GetItemText(value), owner.NativeGetItemText(index), true, CultureInfo.CurrentCulture) != 0)
```

So I tested in my sandbox and found that 
```cs
Console.WriteLine(string.Compare("abc", "abc\0", true, CultureInfo.CurrentCulture));
```

prints out
```
0
```

Essentially meaning that whilst this changed behaviour, it didn't break anything as string.Compare already handles trailing `\0`

That said, I will submit a fix for correctness

**Pull request merged**